File: leetcode/Python/0254-m-factor-combinations.py
"""
DFS
"""
class Solution:
    def getFactors(self, n: int) -> List[List[int]]:
        result = []
        self.helper(result, [], 2, n)
        return result

    def helper(self, result, path, start, remain):
        if remain <= 1:
            if len(path) > 1:
                result.append(path[:])
            return

        import math
        for i in range(start, int(math.sqrt(remain)) + 1):
            if remain % i == 0:
                path.append(i)
                self.helper(result, path, i, remain // i)
                path.pop()

        if remain >= start:
            path.append(remain)
            self.helper(result, path, remain, 1)
            path.pop()

File: leetcode/C++/0254-m-factor-combinations.cpp
class Solution {
private:
    void helper(int start, int n, vector<int>& path, vector<vector<int>>& res) {
        for (int i = start; i <= sqrt(n); i++) {
            if (n % i == 0) {
                path.push_back(i);
                helper(i, n / i, path, res);
                path.push_back(n / i);
                res.push_back(path);
                path.pop_back();
                path.pop_back();
            }
        }
    }
    
public:
    vector<vector<int>> getFactors(int n) {
        vector<int> path;
        vector<vector<int>> res;
        if (n == 1) return res;
        helper(2, n, path, res);
        return res;
    }
};